File: src/orchestrator/observers/debug.py
"""Debug observer for orchestrator event bus.

This module provides the DebugObserver class that subscribes to events and writes
debug files. It replaces the scattered save_* and log_* functions from the old
orchestrator code by centralizing all debug file operations in one place.

Debug files are written to the debug directory with the following naming convention:
- {agent_id}_{state_name}_{step_number:03d}.jsonl - Claude Code streaming output
- {agent_id}_{state_name}_{step_number:03d}.stdout.txt - Script stdout
- {agent_id}_{state_name}_{step_number:03d}.stderr.txt - Script stderr
- {agent_id}_{state_name}_{step_number:03d}.meta.json - Script execution metadata
- transitions.log - State transition history
"""

import json
import logging
from pathlib import Path
from typing import Any, Dict, Optional, TextIO

from ..bus import EventBus
from ..events import (
    ClaudeStreamOutput,
    ScriptOutput,
    StateCompleted,
    StateStarted,
    TransitionOccurred,
)
from ..executors.utils import extract_state_name

logger = logging.getLogger(__name__)


class DebugObserver:
    """Observer that writes debug files based on orchestration events.

    Subscribes to relevant events from the EventBus and writes debug files
    to a specified directory. All file operations are wrapped in try/except
    to ensure failures don't propagate to the orchestration loop.

    The observer maintains open file handles for JSONL streaming output to
    minimize I/O overhead. Call close() when the workflow completes to ensure
    all file handles are properly closed.

    Attributes:
        debug_dir: Directory where debug files are written
    """

    def __init__(self, debug_dir: Path, bus: EventBus) -> None:
        """Initialize the debug observer.

        Args:
            debug_dir: Directory for debug output files
            bus: Event bus to subscribe to
        """
        self.debug_dir = debug_dir
        self._bus = bus

        # Track open file handles for JSONL streaming (agent_id -> file handle)
        self._open_files: Dict[str, TextIO] = {}

        # Track current state name per agent for file naming
        self._agent_states: Dict[str, str] = {}

        # Subscribe to events
        self._subscribe()

    def _subscribe(self) -> None:
        """Subscribe to relevant events on the bus."""
        self._bus.on(StateStarted, self._on_state_started)
        self._bus.on(StateCompleted, self._on_state_completed)
        self._bus.on(ClaudeStreamOutput, self._on_claude_stream_output)
        self._bus.on(ScriptOutput, self._on_script_output)
        self._bus.on(TransitionOccurred, self._on_transition_occurred)

    def _unsubscribe(self) -> None:
        """Unsubscribe from all events on the bus."""
        self._bus.off(StateStarted, self._on_state_started)
        self._bus.off(StateCompleted, self._on_state_completed)
        self._bus.off(ClaudeStreamOutput, self._on_claude_stream_output)
        self._bus.off(ScriptOutput, self._on_script_output)
        self._bus.off(TransitionOccurred, self._on_transition_occurred)

    def close(self) -> None:
        """Clean up resources - close open file handles and unsubscribe.

        Should be called when the workflow completes to ensure all data
        is flushed and file handles are properly closed.
        """
        # Close all open JSONL file handles
        for agent_id, file_handle in self._open_files.items():
            try:
                file_handle.close()
            except Exception as e:
                logger.warning(f"Failed to close file handle for agent {agent_id}: {e}")

        self._open_files.clear()
        self._agent_states.clear()

        # Unsubscribe from events
        self._unsubscribe()

    def _get_jsonl_filepath(
        self,
        agent_id: str,
        state_name: str,
        step_number: int
    ) -> Path:
        """Get the filepath for a Claude Code JSONL output file.

        Args:
            agent_id: Agent identifier
            state_name: State name (without extension)
            step_number: Step number for this agent

        Returns:
            Path to the JSONL file
        """
        filename = f"{agent_id}_{state_name}_{step_number:03d}.jsonl"
        return self.debug_dir / filename

    def _on_state_started(self, event: StateStarted) -> None:
        """Handle StateStarted event - track current state for file naming.

        Args:
            event: StateStarted event
        """
        state_name = extract_state_name(event.state_name)
        self._agent_states[event.agent_id] = state_name

    def _on_state_completed(self, event: StateCompleted) -> None:
        """Handle StateCompleted event - close any open file handles for the agent.

        Args:
            event: StateCompleted event
        """
        # Close JSONL file handle if open
        if event.agent_id in self._open_files:
            try:
                self._open_files[event.agent_id].close()
            except Exception as e:
                logger.warning(
                    f"Failed to close JSONL file for agent {event.agent_id}: {e}"
                )
            del self._open_files[event.agent_id]

    def _on_claude_stream_output(self, event: ClaudeStreamOutput) -> None:
        """Handle ClaudeStreamOutput event - append JSON to JSONL file.

        Uses progressive writes with immediate flush to preserve data on crash.

        Args:
            event: ClaudeStreamOutput event
        """
        state_name = extract_state_name(event.state_name)
        filepath = self._get_jsonl_filepath(
            event.agent_id,
            state_name,
            event.step_number
        )

        try:
            # Get or create file handle for this agent
            if event.agent_id not in self._open_files:
                self._open_files[event.agent_id] = open(
                    filepath, 'a', encoding='utf-8'
                )

            file_handle = self._open_files[event.agent_id]

            # Write JSON line and flush immediately
            file_handle.write(json.dumps(event.json_object, ensure_ascii=False) + '\n')
            file_handle.flush()

        except Exception as e:
            logger.warning(f"Failed to append Claude output to {filepath}: {e}")

    def _on_script_output(self, event: ScriptOutput) -> None:
        """Handle ScriptOutput event - write stdout, stderr, and metadata files.

        Args:
            event: ScriptOutput event
        """
        state_name = extract_state_name(event.state_name)
        base_filename = f"{event.agent_id}_{state_name}_{event.step_number:03d}"

        # Write stdout
        stdout_filepath = self.debug_dir / f"{base_filename}.stdout.txt"
        try:
            with open(stdout_filepath, 'w', encoding='utf-8') as f:
                f.write(event.stdout)
        except Exception as e:
            logger.warning(f"Failed to save script stdout to {stdout_filepath}: {e}")

        # Write stderr
        stderr_filepath = self.debug_dir / f"{base_filename}.stderr.txt"
        try:
            with open(stderr_filepath, 'w', encoding='utf-8') as f:
                f.write(event.stderr)
        except Exception as e:
            logger.warning(f"Failed to save script stderr to {stderr_filepath}: {e}")

        # Write metadata
        metadata_filepath = self.debug_dir / f"{base_filename}.meta.json"
        metadata = {
            "exit_code": event.exit_code,
            "execution_time_ms": event.execution_time_ms,
            "env_vars": event.env_vars
        }
        try:
            with open(metadata_filepath, 'w', encoding='utf-8') as f:
                json.dump(metadata, f, indent=2, ensure_ascii=False)
        except Exception as e:
            logger.warning(f"Failed to save script metadata to {metadata_filepath}: {e}")

    def _on_transition_occurred(self, event: TransitionOccurred) -> None:
        """Handle TransitionOccurred event - append to transitions.log.

        Args:
            event: TransitionOccurred event
        """
        log_file = self.debug_dir / "transitions.log"

        try:
            with open(log_file, 'a', encoding='utf-8') as f:
                # Format log entry
                if event.to_state:
                    f.write(
                        f"{event.timestamp.isoformat()} [{event.agent_id}] "
                        f"{event.from_state} -> {event.to_state} ({event.transition_type})\n"
                    )
                else:
                    f.write(
                        f"{event.timestamp.isoformat()} [{event.agent_id}] "
                        f"{event.from_state} -> (result, terminated)\n"
                    )

                # Write metadata if present
                for key, value in event.metadata.items():
                    f.write(f"  {key}: {value}\n")

                f.write("\n")

        except Exception as e:
            logger.warning(f"Failed to write to transitions.log: {e}")

File: internal/observers/debug/debug.go
// Package debug implements a debug observer that records complete workflow
// execution history to disk for analysis and troubleshooting.
//
// DebugObserver subscribes to ClaudeStreamOutput and TransitionOccurred events
// and writes:
//   - One JSONL file per agent step:
//     {debugDir}/{agentID}_{stateStem}_{step:03d}.jsonl
//   - A chronological state transition log: {debugDir}/transitions.log
//
// The debug directory path is taken from WorkflowStarted.DebugDir, which is
// set by the orchestrator. When DebugDir is empty (debug disabled) the
// observer is a no-op.
//
// All file I/O errors are silently ignored so that debug mode never disrupts
// workflow execution.
package debug

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"

	"github.com/vector76/raymond/internal/bus"
	"github.com/vector76/raymond/internal/events"
)

// DebugObserver writes JSONL step files and a transitions log.
type DebugObserver struct {
	mu       sync.Mutex
	debugDir string // set by WorkflowStarted; empty = no-op
	cancels  []func()
}

// New creates a DebugObserver subscribed to b.
func New(b *bus.Bus) *DebugObserver {
	o := &DebugObserver{}
	o.cancels = []func(){
		bus.Subscribe(b, o.onWorkflowStarted),
		bus.Subscribe(b, o.onClaudeStreamOutput),
		bus.Subscribe(b, o.onTransitionOccurred),
	}
	return o
}

// Close unregisters all subscriptions from the bus.
func (o *DebugObserver) Close() {
	for _, cancel := range o.cancels {
		cancel()
	}
	o.cancels = nil
}

func (o *DebugObserver) onWorkflowStarted(e events.WorkflowStarted) {
	if e.DebugDir == "" {
		return
	}
	_ = os.MkdirAll(e.DebugDir, 0o755)
	o.mu.Lock()
	o.debugDir = e.DebugDir
	o.mu.Unlock()
}

func (o *DebugObserver) onClaudeStreamOutput(e events.ClaudeStreamOutput) {
	o.mu.Lock()
	dir := o.debugDir
	o.mu.Unlock()
	if dir == "" {
		return
	}

	stem := stripExt(e.StateName)
	filename := fmt.Sprintf("%s_%s_%03d.jsonl", e.AgentID, stem, e.StepNumber)
	path := filepath.Join(dir, filename)

	line, err := json.Marshal(e.JSONObject)
	if err != nil {
		return
	}
	appendToFile(path, string(line)+"\n")
}

func (o *DebugObserver) onTransitionOccurred(e events.TransitionOccurred) {
	o.mu.Lock()
	dir := o.debugDir
	o.mu.Unlock()
	if dir == "" {
		return
	}

	logPath := filepath.Join(dir, "transitions.log")
	var sb strings.Builder

	ts := e.Timestamp.Format("2006-01-02T15:04:05.000000")
	if e.ToState != "" {
		fmt.Fprintf(&sb, "%s [%s] %s -> %s (%s)\n",
			ts, e.AgentID, e.FromState, e.ToState, e.TransitionType)
	} else {
		fmt.Fprintf(&sb, "%s [%s] %s -> (result, terminated)\n",
			ts, e.AgentID, e.FromState)
	}

	// Write sorted metadata key-value pairs for deterministic output.
	for _, k := range sortedKeys(e.Metadata) {
		fmt.Fprintf(&sb, "  %s: %v\n", k, e.Metadata[k])
	}
	sb.WriteString("\n")

	appendToFile(logPath, sb.String())
}

// appendToFile appends text to path, creating the file if necessary.
// Errors are silently ignored.
func appendToFile(path, text string) {
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	_, _ = f.WriteString(text)
}

// stripExt removes the last known workflow-state extension from name.
func stripExt(name string) string {
	lower := strings.ToLower(name)
	for _, ext := range []string{".md", ".sh", ".bat", ".ps1"} {
		if strings.HasSuffix(lower, ext) {
			return name[:len(name)-len(ext)]
		}
	}
	return name
}

// sortedKeys returns the keys of m in ascending sorted order.
// Uses simple insertion sort (maps are small).
func sortedKeys(m map[string]any) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	for i := 1; i < len(keys); i++ {
		for j := i; j > 0 && keys[j] < keys[j-1]; j-- {
			keys[j], keys[j-1] = keys[j-1], keys[j]
		}
	}
	return keys
}
